gpexp/importbatch.go

```diff
@@ -0,0 +1,743 @@
+// associated memory.
+func (b *Batch) Import() error {
+	// first we need to translate the toTranslate, then fill out the missing row IDs
+	err := b.doTranslation()
+	if err != nil {
+		return errors.Wrap(err, "doing Translation")
+	}
+
+	// create bitmaps out of each field in b.rowIDs and import. Also
+	// import int data.
+	err = b.doImport()
+	if err != nil {
+		return errors.Wrap(err, "doing import")
+	}
+
+	// clear existing structures without reclaiming the memory
+	b.reset()
+	return nil
+}
+
+func (b *Batch) doTranslation() error {
+	var keys []string
+
+	// translate column keys if there are any
+	if len(b.toTranslateID) > 0 {
+		keys = make([]string, 0, len(b.toTranslateID))
+		for k := range b.toTranslateID {
+			keys = append(keys, k)
+		}
+		ids, err := b.client.TranslateColumnKeys(b.index, keys)
+		if err != nil {
+			return errors.Wrap(err, "translating col keys")
+		}
+		if err := b.transCache.AddCols(b.index.Name(), keys, ids); err != nil {
+			return errors.Wrap(err, "adding cols to cache")
+		}
+		for j, key := range keys {
+			id := ids[j]
+			for _, recordIdx := range b.toTranslateID[key] {
+				b.ids[recordIdx] = id
+			}
+		}
+	} else {
+		keys = make([]string, 0)
+	}
+
+	// translate row keys
+	for i, tt := range b.toTranslate {
+		fieldName := b.header[i].Name()
+		keys = keys[:0]
+
+		// make a slice of keys
+		for k := range tt {
+			keys = append(keys, k)
+		}
+
+		if len(keys) == 0 {
+			continue
+		}
+
+		// translate keys from Pilosa
+		ids, err := b.client.TranslateRowKeys(b.headerMap[fieldName], keys)
+		if err != nil {
+			return errors.Wrap(err, "translating row keys")
+		}
+		if err := b.transCache.AddRows(b.index.Name(), fieldName, keys, ids); err != nil {
+			return errors.Wrap(err, "adding rows to cache")
+		}
+
+		// fill out missing IDs in local batch records with translated IDs
+		rows := b.rowIDs[i]
+		for j, key := range keys {
+			id := ids[j]
+			for _, recordIdx := range tt[key] {
+				rows[recordIdx] = id
+			}
+		}
+	}
+
+	for fieldName, tt := range b.toTranslateSets {
+		keys = keys[:0]
+
+		for k := range tt {
+			keys = append(keys, k)
+		}
+
+		if len(keys) == 0 {
+			continue
+		}
+		// translate keys from Pilosa
+		ids, err := b.client.TranslateRowKeys(b.headerMap[fieldName], keys)
+		if err != nil {
+			return errors.Wrap(err, "translating row keys")
+		}
+		if err := b.transCache.AddRows(b.index.Name(), fieldName, keys, ids); err != nil {
+			return errors.Wrap(err, "adding rows to cache")
+		}
+		rowIDSets := b.rowIDSets[fieldName]
+		for j, key := range keys {
+			rowID := ids[j]
+			for _, recordIdx := range tt[key] {
+				rowIDSets[recordIdx] = append(rowIDSets[recordIdx], rowID)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (b *Batch) doImport() error {
+	eg := errgroup.Group{}
+
+	frags, err := b.makeFragments()
+	if err != nil {
+		return errors.Wrap(err, "making fragments")
+	}
+	for shard, fieldMap := range frags {
+		for field, viewMap := range fieldMap {
```

Comment:
had quite a bit of time to play around with this way of ingesting on our data - maybe this info would be useful for you guys:

this way of ingesting a fixed size batch would heavily bottleneck when number of shards grows and data that is coming to the batch start to be distributed over more and more shards and therefore fragments. Eventually what happened with out data is that doesnt mapper how small batch is configured - we get 1-2 most recent shards in a batch that have most of the data in them, about 70% just in them, they are big. And then we get 100s and 100s of tiny shards. And importing these 100s of tiny shards to Pilosa would heavily bottleneck on unioning bitmaps on the Pilosa cluster, since data read from disk would inevitably happen.

That is what hard limit on batch size does when IDs repeat themselves, how bad it would be depends on the pattern of how IDs repeat themselves in the data, for us it was extremely bad.

What proved to work for us eventually - is to go more granular, put hard limit not to a batch, but to every single fragment. I.e. import fragment to Pilosa as soon as it reaches certain size and flush it from memory. Plus limit number tiny fragment bitmaps in memory by time - to avoid memory blowing up from thousands of tiny fragment bitmaps.

This tremendously reduces and in fact puts an upper bound on number of unions Pilosa cluster has to do during import irrespective of the shard size. It made things a lot faster

Reply:
maybe you can consider implementing streaming ingestion when rows are streamed into go-pilosa API and are flushed to Pilosa cluster periodically when above conditions are met

Reply:
or maybe I can submit my code of this logic to go-pilosa if you guys are interested in something like that

Reply:
@dmibor thanks for the input!

This experiment is definitely skewed toward making the easier cases really fast rather than handling the more difficult scattered/semi-random ID problem. 

We have some Pilosa work going on which is making it more efficient to send small updates to Pilosa, and will have to see how that affects things.

I would be interested to understand how you decided to limit the number of small fragments in memory - you said "by time", but I'm not sure I understand if that's like oldest, or LRU or... 

FWIW, this batch is fixed by number of records, not number of bits in a given shard, so no matter how scattered the IDs are, there is a hard upper limit on memory usage.

Reply:
Yeah, I do see that this way go-pilosa memory footprint would be stable.

"by time" meaning flush small fragments every N minutes, i.e. for every fragment if can be flushed if 1) it's too big and bit count is more than some config, say 1M bits 2) if it is sitting in memory more than N configured minutes - this timeout makes sure tiny fragments sent to Pilosa not too often and are buffered in memory of go-pilosa instead as opposed to fixed batch way of sending when tiny fragments are send as often as batch fills up, multiple times per second at least.

On the other hard, that also puts memory footprint under control, fragment bitmaps can not become too big or there can not be too many of fragments to exhaust memory.